storage/metamorphic: TestPebbleEquivalence failed

storage/metamorphic.TestPebbleEquivalence [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=5675680&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=5675680&tab=artifacts#/) on release-22.1 @ [6dd9fbe94449d8c66ff06adbefb3855a3746665b](https://github.com/cockroachdb/cockroach/commits/6dd9fbe94449d8c66ff06adbefb3855a3746665b):


```
=== RUN   TestPebbleEquivalence
    test_log_scope.go:79: test logs captured to: /artifacts/tmp/_tmp/b2a21e13e3bb63e34c8b8052805e4125/logTestPebbleEquivalence61660856
    test_log_scope.go:80: use -show-logs to present logs inline
first run output file: /artifacts/tmp/_tmp/b2a21e13e3bb63e34c8b8052805e4125/TestPebbleEquivalence3100881545/output.meta
=== CONT  TestPebbleEquivalence
    meta_test.go:174: -- test log scope end --
--- FAIL: TestPebbleEquivalence (66.40s)
=== RUN   TestPebbleEquivalence/engine/standard=7
check file = /artifacts/tmp/_tmp/b2a21e13e3bb63e34c8b8052805e4125/TestPebbleEquivalence3100881545/output.meta
output file = /artifacts/tmp/_tmp/b2a21e13e3bb63e34c8b8052805e4125/TestPebbleEquivalence_engine_standard=72420147106/output.meta
store path = /artifacts/tmp/_tmp/b2a21e13e3bb63e34c8b8052805e4125/TestPebbleEquivalence_engine_standard=71276919892/store
    generator.go:416: mismatching output at line 4210, operation index 4109: expected ok, deleted span = "aptjbtctwcqvgjl" - "gqynlydwqvaru", resumeSpan = <nil>
        , got error: intentIter at intent, but iter not at provisional value
    --- FAIL: TestPebbleEquivalence/engine/standard=7 (0.85s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=bazel,gss,deadlock
</p>
</details>
/cc @cockroachdb/storage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestPebbleEquivalence.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-17383

Epic CRDB-16238

--- a/pkg/storage/metamorphic/BUILD.bazel
+++ b/pkg/storage/metamorphic/BUILD.bazel
@@ -23,6 +23,7 @@ go_library(
         "//pkg/util/syncutil",
         "//pkg/util/uint128",
         "//pkg/util/uuid",
+        "@com_github_cockroachdb_errors//:errors",
         "@com_github_cockroachdb_pebble//:pebble",
         "@com_github_cockroachdb_pebble//vfs",
     ],

--- a/pkg/storage/metamorphic/operations.go
+++ b/pkg/storage/metamorphic/operations.go
@@ -26,6 +26,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/hlc"
 	"github.com/cockroachdb/cockroach/pkg/util/uint128"
 	"github.com/cockroachdb/cockroach/pkg/util/uuid"
+	"github.com/cockroachdb/errors"
 )
 
 // opReference represents one operation; an opGenerator reference as well as
@@ -206,11 +207,17 @@ func (m mvccPutOp) run(ctx context.Context) string {
 	txn.Sequence++
 	writer := m.m.getReadWriter(m.writer)
 
-	err := storage.MVCCPut(ctx, writer, nil, m.key, txn.WriteTimestamp, hlc.ClockTimestamp{}, m.value, txn)
+	err := storage.MVCCPut(ctx, writer, nil, m.key, txn.ReadTimestamp, hlc.ClockTimestamp{}, m.value, txn)
 	if err != nil {
+		if writeTooOldErr := (*roachpb.WriteTooOldError)(nil); errors.As(err, &writeTooOldErr) {
+			txn.WriteTimestamp.Forward(writeTooOldErr.ActualTimestamp)
+			// Update the txn's lock spans to account for this intent being written.
+			addKeyToLockSpans(txn, m.key)
+		}
 		return fmt.Sprintf("error: %s", err)
 	}
 
+	// Update the txn's lock spans to account for this intent being written.
 	addKeyToLockSpans(txn, m.key)
 	return "ok"
 }
@@ -230,11 +237,17 @@ func (m mvccCPutOp) run(ctx context.Context) string {
 	txn.Sequence++
 
 	err := storage.MVCCConditionalPut(ctx, writer, nil, m.key,
-		txn.WriteTimestamp, hlc.ClockTimestamp{}, m.value, m.expVal, true, txn)
+		txn.ReadTimestamp, hlc.ClockTimestamp{}, m.value, m.expVal, true, txn)
 	if err != nil {
+		if writeTooOldErr := (*roachpb.WriteTooOldError)(nil); errors.As(err, &writeTooOldErr) {
+			txn.WriteTimestamp.Forward(writeTooOldErr.ActualTimestamp)
+			// Update the txn's lock spans to account for this intent being written.
+			addKeyToLockSpans(txn, m.key)
+		}
 		return fmt.Sprintf("error: %s", err)
 	}
 
+	// Update the txn's lock spans to account for this intent being written.
 	addKeyToLockSpans(txn, m.key)
 	return "ok"
 }
@@ -252,11 +265,17 @@ func (m mvccInitPutOp) run(ctx context.Context) string {
 	writer := m.m.getReadWriter(m.writer)
 	txn.Sequence++
 
-	err := storage.MVCCInitPut(ctx, writer, nil, m.key, txn.WriteTimestamp, hlc.ClockTimestamp{}, m.value, false, txn)
+	err := storage.MVCCInitPut(ctx, writer, nil, m.key, txn.ReadTimestamp, hlc.ClockTimestamp{}, m.value, false, txn)
 	if err != nil {
+		if writeTooOldErr := (*roachpb.WriteTooOldError)(nil); errors.As(err, &writeTooOldErr) {
+			txn.WriteTimestamp.Forward(writeTooOldErr.ActualTimestamp)
+			// Update the txn's lock spans to account for this intent being written.
+			addKeyToLockSpans(txn, m.key)
+		}
 		return fmt.Sprintf("error: %s", err)
 	}
 
+	// Update the txn's lock spans to account for this intent being written.
 	addKeyToLockSpans(txn, m.key)
 	return "ok"
 }
@@ -333,8 +352,13 @@ func (m mvccDeleteOp) run(ctx context.Context) string {
 	writer := m.m.getReadWriter(m.writer)
 	txn.Sequence++
 
-	err := storage.MVCCDelete(ctx, writer, nil, m.key, txn.WriteTimestamp, hlc.ClockTimestamp{}, txn)
+	err := storage.MVCCDelete(ctx, writer, nil, m.key, txn.ReadTimestamp, hlc.ClockTimestamp{}, txn)
 	if err != nil {
+		if writeTooOldErr := (*roachpb.WriteTooOldError)(nil); errors.As(err, &writeTooOldErr) {
+			txn.WriteTimestamp.Forward(writeTooOldErr.ActualTimestamp)
+			// Update the txn's lock spans to account for this intent being written.
+			addKeyToLockSpans(txn, m.key)
+		}
 		return fmt.Sprintf("error: %s", err)
 	}
 

